### PR TITLE
Add frame pointers support for ARM64 on Linux and macOS

### PR DESCRIPTION
**Pull request by tmcgilchrist**
### Add frame pointers support for ARM64 on Linux and macOS

Frame pointers can be used to walk the stack of function calls in a program. This stack-walking technique is commonly used by external profilers and debuggers, including Linux perf and eBPF, FreeBSD and macOS Dtrace, and GDB and LLDB, to reconstruct the call graph for programs. Enabling frame pointers opens up opportunities for third party debugging and performance tools to understand OCaml programs, without needing OCaml specific changes to understand OCaml's calling conventions.

The OCaml compiler already has an option for `enable-frame-pointers` since 4.14 on amd64 (Originally introduced by @lefessan in #5721, first released in OCaml 4.01 in 2013). In OCaml 5.1 support was added back for Linux in [#11144](https://github.com/ocaml/ocaml/pull/11144) with macOS [#13163](https://github.com/ocaml/ocaml/pull/13163) support coming in 5.3.

This PR extends frame pointer support to ARM64 macOS and Linux. Note sp is the stack pointer register, x29 is the frame pointer register and x30 is the link register. Thanks to @fabbing and @dustanddreams for their assistance.

### Changes

22fee2d Stores sp in temporary register x19 during Iextcall in the same way as amd64 backend. Previously this was implemented in [Save and restore frame pointer across Iextcall on ARM64 #13079](https://github.com/ocaml/ocaml/pull/13079) as push/pop on the stack, however this interferes with frame pointers.

aa2a86d Implements frame pointer save and restore in both the code generation and arm64 runtime. I am using the `stp` and `ldp` instruction plus an extra `add` instruction for updating `x29`.  This results in an extra `add` plus saving an extra register for functions allocating a stack frame. In assembly this typically looks like:

``` assembly
;; prologue
	stp	x29, x30, [sp, #-16]
	sub	sp, sp, #16
	add	x29,  sp, #0

;; epilogue
	add	sp, sp, #16
	ldp	x29, x30, [sp, #-16]
```

The compiler already maintains `x29` register when calling into C and when using TSan. There is minimal extra stack space usage, as we needed to be quad word aligned anyway, so the `x29` is going into an already allocated space on the stack. I borrowed the same `UPDATE_BASE_POINTER` and `Lentertrap` tricks from https://github.com/ocaml/ocaml/pull/11144 and https://github.com/ocaml/ocaml/pull/11031.

fea28d6 and caccae2 work around inlining differences across platforms and add a simple backtrace tests for exceptions, which I felt was missing previously.

18ad55c Enables frame-pointers in CI to demonstrate the test suite passes but I expect frame pointer support in GitHub CI to be done differently.

I haven't done comprehensive performance testing of this change across Linux and macOS, however I would expect the impact to be similar to x86_64. If someone would like some hard numbers, I can run some tests.

### Background

It's interesting to note that the [ABI on macOS](https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms) requires maintaining frame pointers, and certain popular Linux distributions like [Ubuntu](https://ubuntu.com/blog/ubuntu-performance-engineering-with-frame-pointers-by-default), [Fedora](https://pagure.io/fesco/issue/2923) and [Arch](https://gitlab.archlinux.org/archlinux/rfcs/-/merge_requests/26) are reenabling frame pointers in recent distributions. 

### Fixing perf profiling
As a further demonstration that this fixes issues with misattributed frames in perf 
(See [#12563](https://github.com/ocaml/ocaml/issues/12563) for a longer discussion on the topic).

Using this test program compiled with `ocamlopt sched.ml -o sched.exe` in a switch with this change on Ubuntu 24.04 LTS (which is compiled with frame pointers turned on):
```ocaml
open Effect
open Effect.Deep

exception E
type _ t += Yield : unit t
          | Fork : (unit -> string) -> unit t
          | Ping : unit t
exception Pong

let say _ =
  (* Printf.printf "%s%!" x *)
  ignore(Sys.opaque_identity(Array.make 1_000 0))

let run main =
  let run_q = Queue.create () in
  let enqueue k = Queue.push k run_q in
  let rec dequeue () =
    if Queue.is_empty run_q then `Finished
    else continue (Queue.pop run_q) ()
  in
  let rec spawn f =
    let[@inline never] spawn_retc = function
      | "ok" -> say "."; dequeue ()
      | s -> failwith ("Unexpected result: " ^ s)
    in
    let[@inline never] spawn_exnc = function
      | E -> say "!"; dequeue ()
      | e -> raise e
    in
    let[@inline never] spawn_effc (type a) (e : a t) =
      match e with
        | Yield -> Some (fun (k : (a, _) continuation) ->
            say ","; enqueue k; dequeue ())
        | Fork f -> Some (fun (k : (a, _) continuation) ->
            say "+"; enqueue k; spawn f)
        | Ping -> Some (fun (k : (a, _) continuation) ->
            say "["; discontinue k Pong)
        | _ -> None
    in
    match_with f ()
    { retc = spawn_retc;
      exnc = spawn_exnc;
      effc = spawn_effc }
  in
  spawn main

let[@inline never] g () =
  let[@inline never] g_comp () = perform Ping; failwith "no pong" in
  let[@inline never] g_retc x = x in
  let[@inline never] g_exnc = function | Pong -> say "]" | e -> raise e in
  let[@inline never] g_effc (type a) (e : a t) =
    match e with
    | Yield -> Some (fun (k : (a,_) continuation) -> failwith "what?")
    | _ -> None in
  perform Yield; say "C"; perform Yield;
  begin match_with g_comp ()
  { retc = g_retc;
    exnc = g_exnc;
    effc = g_effc }
  end;
  raise E

let[@inline never] h () = say "B"; "ok"

let[@inline never] test () =
  say "A";
  perform (Fork g);
  perform (Fork h);
  say "D";
  perform Yield;
  say "E";
  "ok"

let () =
  for a = 1 to 100_000 do
  let `Finished = run test in
  say "\n"
  done
```
First using `sudo perf record --call-graph dwarf -o sched-perf-dwarf.data ./sched.exe`, note `cfree` pointing to an address.
Screenshot 2024-09-25 at 2 57 34 PM

Now with `sudo perf record --call-graph fp -o sched-perf-fp.data ./sched.exe` the `cfree` node is correctly included.
Screenshot 2024-09-25 at 2 57 17 PM

### CODE REVIEW

**Comment by gasche** [comment]
(Nitpick: frame-pointers support was introduced by @lefessan in #5721, first released in OCaml 4.01 in 2013.)

File: asmcomp/arm64/emit.mlp
Comment (ghost) [inline asmcomp/arm64/emit.mlp]:
```suggestion
      `	ldp	x29, x30, [sp, #-16]\n`;
```

File: asmcomp/arm64/emit.mlp
Comment (ghost) [inline asmcomp/arm64/emit.mlp]:
```suggestion
      `	ldr	x30, [sp, #-8]\n`;
```

**Review by ghost — APPROVED** [review]

**Comment by kayceesrk** [comment]
> and GDB and LLDB, to reconstruct the call graph for programs.

IIUC, GDB and LLDB default to DWARF and may optionally use the frame pointer info to reconstruct call graph, right?

File: asmcomp/arm64/emit.mlp
Comment (kayceesrk) [inline asmcomp/arm64/emit.mlp:90]: Nit: can the "optional frame pointer" comment be attached to the inner conditional expression?

Comment (kayceesrk) [inline asmcomp/arm64/emit.mlp]: Nit: can the "optional frame pointer" comment be attached to the inner conditional expression?

File: runtime/arm64.S
Comment (kayceesrk) [inline runtime/arm64.S]:
```suggestion
   frame-pointers if it has been split (i.e., in a continuation).
```

**Comment by kayceesrk** [comment]
```
;; prologue
	stp	x29, x30, [sp, #-16]
	sub	sp, sp, #16
	add	x29,  sp, #0

;; epilogue
	add	sp, sp, #16
	ldp	x29, x30, [sp, #-16]
```

does not match the [prologue code](https://github.com/ocaml/ocaml/pull/13500/files#diff-8cf9769d53db2ce59142888a963afd141b3bc0dc522888ef3f53efbb6479542aR686) which is an `stp` followed by an `add` instruction. It would be useful to update the description to match the code.

**Review by kayceesrk — APPROVED** [review]

**Review by fabbing — APPROVED** [review]

**Comment by tmcgilchrist** [comment]
> > and GDB and LLDB, to reconstruct the call graph for programs.
> 
> IIUC, GDB and LLDB default to DWARF and may optionally use the frame pointer info to reconstruct call graph, right?

That is correct. 

For GDB you can control which unwinders get used with `maintenance set dwarf unwinders [off/on]` and `maintenance show dwarf unwinders` to see if they are enabled. So you can build something with frame pointers, load it into GDB and then turn off the DWARF CFI based unwinding to just use frame pointers.

For LLDB I haven't found a similar command. You can run `log enable lldb unwind` to turn on verbose logging of the backtrace calculations but that will always use DWARF CFI since OCaml always outputs that information.

**Comment by kayceesrk** [comment]
Thanks. Can you please correct the original description?

**Comment by sidkshatriya** [comment]
> For GDB you can control which unwinders get used with maintenance set dwarf unwinders [off/on] and maintenance show dwarf unwinders to see if they are enabled.

If you enable frame pointers support using this PR (when using OCaml) or enable frame pointers support using (`-fno-omit-frame-pointers`) when using C/C++ does it matter what unwinder is being used via `maintenance set dwarf unwinders [on/off]` ?

My understanding is that the DWARF information embedded in the executables would simply describe each frame uses the frame pointer. So in other words telling gdb to switch on/off the unwinder should make any difference. (Here, I notice that you're adding `.cfi_` directives for OCaml -- they should be translated to the appropriate DWARF unwind information by the assembler).

I guess a difference would only arise if the DWARF information was stripped from the executable. In that scenario it would useful to tell gdb that even if there was no DWARF unwind information found, it should use frame pointers.

**Comment by tmcgilchrist** [comment]
@kayceesrk I think you might be overlooking the `emit_stack_adjustment` call which adjusts `sp` giving the `sub sp, sp, #16`. https://github.com/ocaml/ocaml/pull/13500/files#diff-8cf9769d53db2ce59142888a963afd141b3bc0dc522888ef3f53efbb6479542aR695

The order of these instructions was changed so we could use the `stp/ldp` instructions. Some OCaml code allocates large frames and I couldn't encode the larger offsets in those instructions without changing the order. @dustanddreams suggested changing the emit.mlp for prologue/epilogue to avoid these errors.

Currently OCaml compiler would generate them in the opposite order:

```
;; prologue
	sub	sp, sp, #16
	str	x30, [sp, #-8]

;; epilogue
	ldr	x30, [sp, #-8]
	add	sp, sp, #16
```

**Comment by kayceesrk** [comment]
> Thanks. Can you please correct the original description?

My intention was to avoid making the claim that frame pointers were necessary for GDB and LLDB backtraces. I reread the description and it says "commonly used by.." which seems ok. No correction is necessary.

**Comment by kayceesrk** [comment]
>  I think you might be overlooking the emit_stack_adjustment call which adjusts sp giving the sub sp, sp, #16.

Ok.

**Comment by kayceesrk** [comment]
The PR looks good to me, and we've got 3 approvals. Unless anyone objects, I plan to merge this after the Changes file conflict is resolved.

**Comment by tmcgilchrist** [comment]
> If you enable frame pointers support using this PR (when using OCaml) or enable frame pointers support using (`-fno-omit-frame-pointers`) when using C/C++ does it matter what unwinder is being used via `maintenance set dwarf unwinders [on/off]` ?
 
This PR only changes the OCaml generated code and runtime to maintain frame pointers, we already maintain frame pointers when calling C/C++ code. Whether that C/C++ code maintains frame pointers depends on the Linux distribution you use.  The use of frame pointers for debugging is one possibility, in that you could use it if necessary, but the bigger motivating use is for perf #12563 
 
> My understanding is that the DWARF information embedded in the executables would simply describe each frame uses the frame pointer. So in other words telling gdb to switch on/off the unwinder should make any difference. (Here, I notice that you're adding `.cfi_` directives for OCaml -- they should be translated to the appropriate DWARF unwind information by the assembler).
> 
> I guess a difference would only arise if the DWARF information was stripped from the executable. In that scenario it would useful to tell gdb that even if there was no DWARF unwind information found, it should use frame pointers.

That's not quite correct, OCaml will already generate accurate CFI information but doesn't maintain x29 for OCaml code. This PR is adding that second part. 

I haven't tested what will happen if you strip an OCaml executable in that case. I would expect that GDB would use frame pointers. The [GDB manual talks](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Maintenance-Commands.html) about a priority order for which unwinders it uses.

**Comment by sidkshatriya** [comment]
@tmcgilchrist
After the merge of this PR to the codebase, you have added `.cfi` directives to describe the stack frame in terms of the frame pointer. These `.cfi` directives will cause the appropriate DWARF information to be written to this executable by the assembler. So it should _not_ be necessary to tell GDB to change the unwinder when debugging an OCaml executable compiled with frame pointers. When gdb uses the DWARF unwinder, it will simply use the frame pointer because that's what the `cfi` directives had encoded !

The limited point that I'm making is that the DWARF unwinder can encode very complex unwinding or simple unwinding in terms of the frame pointer. Here the DWARF unwinding section will be simple and trivial because frame pointers are used in each OCaml function and more complex unwinding techniques will not be required.

Yes, things will be better with perf because for performance reasons, I don't think perf tries look at DWARF for unwinding purposes. But if the unwinding is in terms of the frame pointer it becomes trivial for perf to do it by itself.

Also if you stripped DWARF information out of this executable, it should be _still_ be possible to do unwinding in gdb because again unwinding when the frame pointer is used is very easy.

**Comment by tmcgilchrist** [comment]
Thank you for the questions @sidkshatriya my understanding of these things is as follows.

> So it should not be necessary to tell GDB to change the unwinder when debugging an OCaml executable compiled with frame pointers.

Perhaps there is a misunderstanding here. I'm only mentioning the GDB command for turning on/off dwarf unwinders so reviewers could verify that the frame pointers are accurate. For someone using GDB on OCaml programs, it will use the DWARF CFI unwinding as expected.


> Yes, things will be better with perf because for performance reasons,
> I don't think perf tries look at DWARF for unwinding purposes. But if
> the unwinding is in terms of the frame pointer it becomes trivial for
> perf to do it by itself.

The motivation for using frame pointers with OCaml / perf is less about performance (though that is faster and more compact with frame pointers) and more about the stack layout that OCaml 5 uses. OCaml 5 (aka multicore) introduced non-contiguous stacks as part of the implementation of effects. Mentioned in the PLDI 2021 paper on retrofitting effect handlers -- Section 5.5 in https://kcsrk.info/papers/retro-concurrency_pldi_21.pdf. The non-contiguous stacks do not work accurately with the copying nature of perf, so traces produced for OCaml 5 will be truncated if used with DWARF call graphs.  This copying behaviour is reasonably well-known. Some references:

 * https://rwmj.wordpress.com/2023/02/14/frame-pointers-vs-dwarf-my-verdict/
 * https://dl.acm.org/doi/pdf/10.1145/3360572

> Also if you stripped DWARF information out of this executable, it should be still be possible to do unwinding in gdb because again unwinding when the frame pointer is used is very easy.

I would expect so too, but I haven't looked into this in great detail.

**Pull request merged**
